File: apps/web/lib/chat-api.ts
import { api } from './api';

export interface ChatMessage {
  id: string;
  threadId: string;
  senderId: string;
  body: string;
  sentAt: string;
}

export interface ChatThread {
  id: string;
  clubId: string;
  coachId: string;
  playerId: string;
  coachLastReadAt: string | null;
  playerLastReadAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface ChatThreadSummary extends ChatThread {
  otherPartyName: string;
  otherPartyRole: 'coach' | 'player';
  lastMessage: ChatMessage | null;
  unreadCount: number;
}

export const chatApi = {
  listThreads: () => api<ChatThreadSummary[]>('/chat/threads'),
  startThreadWithPlayer: (playerId: string) =>
    api<ChatThread>('/chat/threads', { method: 'POST', body: { playerId } }),
  listMessages: (threadId: string, opts: { before?: string; limit?: number } = {}) => {
    const params = new URLSearchParams();
    if (opts.before) params.set('before', opts.before);
    if (opts.limit) params.set('limit', String(opts.limit));
    const qs = params.toString();
    return api<ChatMessage[]>(`/chat/threads/${threadId}/messages${qs ? `?${qs}` : ''}`);
  },
  sendMessage: (threadId: string, body: string) =>
    api<ChatMessage>(`/chat/threads/${threadId}/messages`, {
      method: 'POST',
      body: { body },
    }),
  markRead: (threadId: string) =>
    api<void>(`/chat/threads/${threadId}/read`, { method: 'POST' }),
};

File: apps/mobile/lib/api/chat_api.dart
import 'package:dio/dio.dart';
import '../models/chat.dart';
import 'api_client.dart';
import 'api_exception.dart';

class ChatApi {
  final ApiClient client;
  ChatApi(this.client);

  Future<List<ChatThreadSummary>> listThreads() async {
    try {
      final res = await client.dio.get('/chat/threads');
      ensureOk(res);
      final list = res.data as List<dynamic>;
      final items = list
          .map((e) => ChatThreadSummary.fromJson(e as Map<String, dynamic>))
          .toList();
      items.sort((a, b) => b.updatedAt.compareTo(a.updatedAt));
      return items;
    } on DioException catch (e) {
      throw toApiException(e);
    }
  }

  Future<ChatThread> startThreadWithPlayer(String playerId) async {
    try {
      final res = await client.dio.post('/chat/threads', data: {'playerId': playerId});
      ensureOk(res);
      return ChatThread.fromJson(res.data as Map<String, dynamic>);
    } on DioException catch (e) {
      throw toApiException(e);
    }
  }

  Future<List<ChatMessage>> listMessages(String threadId, {DateTime? before, int limit = 50}) async {
    try {
      final res = await client.dio.get('/chat/threads/$threadId/messages',
          queryParameters: {
            if (before != null) 'before': before.toIso8601String(),
            'limit': limit,
          });
      ensureOk(res);
      final list = res.data as List<dynamic>;
      final items = list
          .map((e) => ChatMessage.fromJson(e as Map<String, dynamic>))
          .toList();
      items.sort((a, b) => a.sentAt.compareTo(b.sentAt));
      return items;
    } on DioException catch (e) {
      throw toApiException(e);
    }
  }

  Future<ChatMessage> sendMessage(String threadId, String body) async {
    try {
      final res = await client.dio
          .post('/chat/threads/$threadId/messages', data: {'body': body});
      ensureOk(res);
      return ChatMessage.fromJson(res.data as Map<String, dynamic>);
    } on DioException catch (e) {
      throw toApiException(e);
    }
  }

  Future<void> markRead(String threadId) async {
    try {
      final res = await client.dio.post('/chat/threads/$threadId/read');
      ensureOk(res);
    } on DioException catch (e) {
      throw toApiException(e);
    }
  }
}
